keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
+resource assignments before starting a pod. In future steps kubelet will patch
+this information into node status.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Information about the discovered resources
++    ResourcesInfo resources = 2;
++}
+
++// ResourcesInfo contains information about the resources discovered by the
++// runtime.
++message ResourcesInfo {
++    // Pod-level QoS resources available.
++    repeated QoSResourceInfo pod_qos_resources = 1;
++    // Container-level QoS resources available.
++    repeated QoSResourceInfo container_qos_resources = 2;
++}
+
++// QoSResourceInfo contains information about one type of QoS resource.
++message QoSResourceInfo {
++    // Name of the QoS resources.
++    string Name = 1;
++    // Mutable is set to true if this resource supports in-place updates i.e.
++    // the class of a running container or sandbox can be changed.
++    bool Mutable = 2;
++    // List of classes of this QoS resource.
++    repeated QoSResourceClassInfo classes = 3;
++}
+
++// QoSResourceClassInfo contains information about one class of certain
++// QoS resource.
++message QoSResourceClassInfo {
++    // Name of the class
++    string name = 1;
++    // Capacity is the number of maximum allowed simultaneous assignments into this class
++    // Zero means "infinite" capacity i.e. the usage is not restricted.
++    uint64 capacity = 2;
+ }
+```
+
+#### Consts
+
+Also, define "known" QoS-class resource types to more easily align container
+runtime implementations:
+
+```diff
++const (
++       // QoSResourceRdt is the name of the RDT QoS-class resource
++       QoSResourceRdt = "rdt"
++       // QoSResourceBlockio is the name of the blockio QoS-class resource
++       QoSResourceBlockio = "blockio"
++)
+```
+
+### Kubernetes API
+
+The PodSpec will be extended to support assignment of pod-level and
+container-level QoS-class devices. NodeStatus will be extended to include
+information about the available QoS-class resources on a node.
+
+#### PodSpec
+
+We introduce a new field, QoSResources into the existing ResourceRequirements
+struct. This will enable the assignment of QoS resources for containers.
+
+```diff
+ type ResourceRequirements struct {
+        // Limits describes the maximum amount of compute resources allowed.
+@@ -2195,6 +2198,10 @@ type ResourceRequirements struct {
+        // +featureGate=DynamicResourceAllocation
+        // +optional
+        Claims []ResourceClaim
++       // QoSResources specifies the QoS resources.
++       // +optional
++       QoSResources map[QoSResourceName]string
+ }
+
++// QoSResourceName is the name of a QoS resource.
++type QoSResourceName string
+```
+
+Also, we add a Resources field to the PodSpec, to enable assignment of
+pod-level QoS resources. We will re-use the existing ResourceRequirements type
```

Comment:
Let's not re-use types when we don't have to.  Make it a subset of the container schema, so we can align with pod-level resources if/when they happen, but don't add fields that users can't use.

Reply:
> Let's not re-use types when we don't have to. Make it a subset of the container schema, so we can align with pod-level resources if/when they happen, but don't add fields that users can't use.

Good comment 👍 Any suggestions here? What I was thinking how to align with the recent DRA claims field in the pod spec. I.e. similarly propose a plain `qosResources` field in the PodSpec or specify a separate `PodResourceRequirements` struct to include possible other types of future pod resources?

Reply:
I would define 2 structs with the same (or almost the same) fields - one for Pod and one for Container.  It makes them a LITTLE harder to consume in code but we have been burned so many times by re-using types that I think it's better not to at all.

Reply:
K. In the proposal it was basically just a simple map[string]string (to sort of align with the old resource limits and requests)

```golang
type QoSResourceRequest struct {
        Name string
}
```

and something like this for the ResourceRequirements

```diff
+       QoSResources []QoSResourceRequest
 }
```

and ditto for pod-level resources

Reply:
@thockin this was changed as suggested

In Container:

```diff
 type ResourceRequirements struct {
...
        Claims []ResourceClaim
+       // QOSResources specifies the requested QoS resources.
+       // +optional
+       QOSResources []QOSResourceRequest
 }

+// QOSResourceRequest specifies a request for one QoS resource type.
+type QOSResourceRequest struct {
+       // Name of the QoS resource.
+       Name QOSResourceName
+       // Name of the class (inside the QoS resource type specified by Name field).
+       Class string
+}
```

And in PodSpec as a new field

```diff
 type PodSpec struct {
...
        ResourceClaims []PodResourceClaim
+       QOSResources []PodQOSResourceRequest
 }

+// PodQOSResourceRequest specifies a request for one QoS resource type for a
+// Pod.
+type PodQOSResourceRequest struct {
+       // Name of the QoS resource.
+       Name QOSResourceName
+       // Name of the class (inside the QoS resource type specified by Name field).
+       Class string
+}